plugins/hls-explicit-record-fields-plugin/src/Ide/Plugin/ExplicitFields.hs

```diff
@@ -0,0 +1,266 @@
+                                                  WorkspaceEdit (WorkspaceEdit),
+                                                  fromNormalizedUri,
+                                                  normalizedFilePathToUri,
+                                                  type (|?) (InR))
+import qualified Language.LSP.Types.Lens         as L
+
+
+data Log = LogShake Shake.Log
+
+instance Pretty Log where
+  pretty = \case
+    LogShake shakeLog -> pretty shakeLog
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId = (defaultPluginDescriptor plId)
+  { pluginHandlers = mkPluginHandler STextDocumentCodeAction codeActionProvider
+  , pluginRules = collectRecordsRule recorder
+  }
+
+codeActionProvider :: PluginMethodHandler IdeState 'TextDocumentCodeAction
+codeActionProvider ideState pId (CodeActionParams _ _ docId range _) = pluginResponse $ do
+  nfp <- getNormalizedFilePath (docId ^. L.uri)
+  pragma <- getFirstPragma pId ideState nfp
+  CRR renderedRecs (map unExt -> exts) <- collectRecordsInRange range ideState nfp
+  let actions = map (mkCodeAction nfp exts pragma) renderedRecs
+  pure $ List actions
+
+  where
+    mkCodeAction :: NormalizedFilePath -> [Extension] -> NextPragmaInfo -> RenderedRecordInfo -> Command |? CodeAction
+    mkCodeAction nfp exts pragma rec = InR CodeAction
+      { _title = mkCodeActionTitle exts
+      , _kind = Just CodeActionRefactorRewrite
+      , _diagnostics = Nothing
+      , _isPreferred = Nothing
+      , _disabled = Nothing
+      , _edit = Just $ mkWorkspaceEdit nfp edits
+      , _command = Nothing
+      , _xdata = Nothing
+      }
+      where
+        edits = catMaybes [ mkTextEdit rec , pragmaEdit ]
+
+        mkTextEdit :: RenderedRecordInfo -> Maybe TextEdit
+        mkTextEdit (RenderedRecordInfo ss r) = TextEdit <$> srcSpanToRange ss <*> pure r
+
+        -- NOTE(ozkutuk): `RecordPuns` extension is renamed to `NamedFieldPuns`
+        -- in GHC 9.4, but we still want to insert `NamedFieldPuns` in pre-9.4
+        -- GHC as well, hence the replacement.
+        -- https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6156
+        pragmaEdit :: Maybe TextEdit
+        pragmaEdit = if NamedFieldPuns `elem` exts
+                       then Nothing
+                       else Just $ patchExtName $ insertNewPragma pragma NamedFieldPuns
+          where
+            patchExtName = L.newText %~ T.replace "Record" "NamedField"
+
+    mkWorkspaceEdit :: NormalizedFilePath -> [TextEdit] -> WorkspaceEdit
+    mkWorkspaceEdit nfp edits = WorkspaceEdit changes Nothing Nothing
+      where
+        changes = Just $ HashMap.singleton (fromNormalizedUri (normalizedFilePathToUri nfp)) (List edits)
+
+    mkCodeActionTitle :: [Extension] -> Text
+    mkCodeActionTitle exts =
+      if NamedFieldPuns `elem` exts
+        then title
+        else title <> " (needs extension: NamedFieldPuns)"
```

Comment:
Cute, but I think this is inconsistent UX: we never normally say in the code action title if we're going to insert pragmas. I think the right way to do this consistently is `AnnotatedTextEdit`s : https://github.com/haskell/haskell-language-server/issues/3210

Reply:
I have adapted this from the alternate number format plugin, as seen [here](https://github.com/haskell/haskell-language-server/blob/master/plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/AlternateNumberFormat.hs#L124-L130). I do agree that `AnnotatedTextEdit`s seem to be a nicer way to do this though. I will look into it to see if we can make use of it here.

Reply:
No pressure to use it in this PR - AFAIK we use it nowhere yet. But I'd be delighted if someone looked into it!

Reply:
I did some testing by converting this to use `AnnotatedTextEdit`. It looks like it is a pretty new addition to spec as the editors I tried didn't properly support them. Neovim LSP didn't care, and just handled it like a usual `TextEdit` without the `ChangeAnnotation` niceties. VSCode didn't display the code action at all.

The spec states that servers shouldn't send `AnnotatedTextEdit`s if the client doesn't support them. Therefore wherever we use them, we need to query the client capabilities and fallback to plain `TextEdit`s if needed. Adding some convenience functions to take care of these kinds of details sounds like a good idea, but seems out-of-scope for this PR, so I am leaving this as is.

Reply:
Hmm, that's disappointing, I would have expected vscode to do something :thinking:

Reply:
related: https://github.com/microsoft/vscode/issues/151240